### ISSUE TITLE
Pytorch-dp issue: Train differentially private LSTM for text generation?

### ISSUE THREAD

**Issue by AIessa** [pytorch/opacus#38]
**I would like to integrate pytorch-dp in an LSTM to generate differentially private synthetic text**. Without DP, I was using word-level language modeling RNN (with multi-layer LSTM) from the pytorch/examples repository (https://github.com/pytorch/examples/tree/master/word_language_model). 

I tried a few things to integrate pytorch-dp, but ran into compatibility issues of privacy engine with some of the modules and would like to know whether (and if so, how) the current version could be used in my situation at all.

**If there is a chance that pytorch-dp could be used in this case**, this is what I tried so far:

1. I first followed the readme and created the Privacy Engine, attaching it to an optimizer and the current model. This resulted in an error:
   _“torchdp.dp_model_inspector.IncompatibleModuleException: Model contains incompatible modules. Some modules are not valid.: ['Main.encoder', 'Main.rnn’]”_
 My understanding is that pytorch-dp does not support pytorch RNN and specifically LSTM layers and does not work with the encoder module.

2. You provide a DPLSTM class which theoretically seems like an alternative to the pytorch LSTM model, but seems to be missing functions and properties that are necessary to integrate it in the text generator (e.g. encoder and decoder, but potentially also the possibility of using a multi-layer LSTM or dropout).
  Disregarding the usefulness of the model for the task, using DPLSTM as model still resulted in an issue, but this time the invalid module was just ‘Main’ in general, which I was not sure how to interpret.

I would love to integrate pytorch-dp in my project, but I am aware that it is still in development and the examples seem to be focussed on classification rather than generative models. 

**It would be great if you could let me know whether the issue is with pytorch-dp or my understanding of it.**

Thanks in advance!

**Comment by sayanghosh** [comment]
Hi @AIessa thanks for trying out the LSTM with Opacus. To answer your questions :
1. The layer nn.LSTM by itself is not supported in Opacus. We recommend switching it out with a DPLSTM layer, which is a new addition we have to support LSTM training with DP.
2. Yes, we at present support DPLSTM only for classification purposes, and will be extending to other use cases such as seq2seq or LMs pretty soon. Could you share the code you have which produced the error of invalid module being just 'Main' ?
We have plans to support multi-layer LSTMs and dropout in the future, and you can refer to this for a list of existing features the DPLSTM does support : 
1. A bias term must be present (bias flag set to True)
2. Single directional (bidirectional set to False)
3. Single layer (num_layers set to 1)
4. No dropout supported yet
5. Initial LSTM states (h and c) set to zero
Thanks, and let us know if you have any more questions or wish to contribute.

**Comment by Darktex** [comment]
I think we can close this one. @Alessa please feel free to file another issue if you have further questions we can answer!
